File: Python/ts/serialization/geometry.ts
import { Polygon, Polyline, Vector } from "@/Core/geometry";
import { Json } from "@/Core/utils";
import { destringify_f64_array, stringify_f64_array } from "./number_array";

export function serialize_polyline(l: Polyline): {
    type: "polyline";
    data: Json;
} {
    return {
        type: "polyline",
        data: stringify_f64_array(l.positions)
    };
}

export function serialize_polygon(l: Polygon): {
    type: "polygon";
    data: Json;
} {
    return {
        type: "polygon",
        data: stringify_f64_array(l.positions)
    };
}

export function serialize_vector(v: Vector): {
    type: "vector";
    data: Json;
} {
    return {
        type: "vector",
        data: [v.x, v.y]
    };
}

export function deserialize_polyline(value: {
    type: "polyline";
    data: any;
}): Polyline {
    return new Polyline(destringify_f64_array(value.data));
}

export function deserialize_polygon(value: {
    type: "polygon";
    data: any;
}): Polygon {
    return new Polygon(destringify_f64_array(value.data));
}

export function deserialize_vector(value: {
    type: "vector";
    data: any;
}): Vector {
    const [x, y] = value.data as [number, number];
    return new Vector(x, y);
}

File: Python/stoff_types/src/stoff_types/serialization/geometry.py
from typing import Any, Literal, TypedDict

from .index import Matrix, Polygon, Polyline, Vector
from .number_array import destringify_f64_array, stringify_f64_array


class SerializedPolyline(TypedDict):
    type: Literal["polyline"]
    data: Any


class SerializedPolygon(TypedDict):
    type: Literal["polygon"]
    data: Any


class SerializedVector(TypedDict):
    type: Literal["vector"]
    data: Any


class SerializedMatrix(TypedDict):
    type: Literal["matrix"]
    data: Any


def serialize_polyline(line: Polyline) -> SerializedPolyline:
    return {"type": "polyline", "data": stringify_f64_array(line.points)}


def serialize_polygon(polygon: Polygon) -> SerializedPolygon:
    return {"type": "polygon", "data": stringify_f64_array(polygon.points)}


def serialize_vector(vector: Vector) -> SerializedVector:
    return {"type": "vector", "data": [vector.x, vector.y]}


def serialize_matrix(matrix: Matrix) -> SerializedMatrix:
    return {
        "type": "matrix",
        "data": [matrix.a, matrix.b, matrix.c, matrix.d],
    }


def deserialize_polyline(value: SerializedPolyline) -> Polyline:
    return Polyline(destringify_f64_array(value["data"]))


def deserialize_polygon(value: SerializedPolygon) -> Polygon:
    return Polygon(destringify_f64_array(value["data"]))


def deserialize_vector(value: SerializedVector) -> Vector:
    x, y = value["data"]
    return Vector(x, y)


def deserialize_matrix(value: SerializedMatrix) -> Matrix:
    a, b, c, d = value["data"]
    return Matrix(a, b, c, d)
